libraries/Microsoft.Bot.Connector/Authentication/EndorsementsValidator.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public static class EndorsementsValidator
+    {
+        //Original Code from V3 SDK, shown here
+        //public static bool Validate(IEnumerable<IActivity> activities, string[] endorsements)
+        //{
+        //    return !activities.Select(activity => activity.ChannelId).Except(endorsements).Any();
+        //}
+
+        /// <summary>
+        /// Verify that the set of ChannelIds, which come from the incoming activities,
+        /// all match the endorsements found on the JWT Token. 
+        /// For example, if an Activity comes from DirectLine, that channelId should
+        /// say that, and the jwt token endorsement must also match that. 
+        /// </summary>
+        /// <param name="channelIds"></param>
+        /// <param name="endorsements"></param>
+        /// <returns></returns>
+        public static bool Validate(string channelId, string[] endorsements, bool allowUnendorsedChannels = false)
```

Comment:
So, I didn't want to incorporate this in my original PR because I couldn't tell how the endorsements array was going to be used at the time, but, now that we have figure that out and you're implementing it here, I think that we should optimize this a bit. 

Specifically scanning through the `string[]` with `.Contains` is `O(N)` and I think there's something like 30(?) endorsements in the real world bot service metadata document.  We would be better served performance wise by changing `EndorsementsRetriever` to build/cache `IDictionary<string, HashSet<string>>` and then changing this method to take `HashSet<string>` (or even just `ICollection<string>`) instead of `string[]` so that the `.Contains` check can be backed by the `O(1)` implementation instead.

Reply:
If you agree, I can apply those changes on top of this PR, just LMK.